### PR TITLE
Tweak deserialization code

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: #

### Changelog

* Abstract repeat code in deserialzation.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/serialization/deserialize-dynamic.ts [hunk 1]
```diff
@@ -208,16 +199,9 @@ function compileDeserializeNative (_self: _Deserializer, klass: CCClassConstruct
...
-                } else {
-                    const defaultType = typeof defaultValue;
-                    isPrimitiveType = defaultType === 'string'
-                                      || defaultType === 'number'
-                                      || defaultType === 'boolean';
-                }
+                isPrimitiveType = isPrimitiveTypeInFastMode(defaultValue, userType);
             }
-            if (fastMode && isPrimitiveType) {
+            if (isPrimitiveType) {
```
Comment (shrinktofit) [inline cocos/serialization/deserialize-dynamic.ts]: Note, **according to original code**, `isPrimitive === true` has already meant `fastMode === true`.

Comment (jareguo) [reply cocos/serialization/deserialize-dynamic.ts]: Not necessarily, user-defined fields could also be primitive?

**Review by jareguo — APPROVED** [review]

File: cocos/serialization/deserialize-dynamic.ts [hunk 2]
```diff
@@ -136,17 +136,8 @@ function compileDeserializeJIT (self: _Deserializer, klass: CCClassConstructor<u
...
-                isPrimitiveType = userType instanceof CCClass.Attr.PrimitiveType || userType === ENUM_TAG || userType === BITMASK_TAG;
-            } else {
-                const defaultType = typeof defaultValue;
-                isPrimitiveType = defaultType === 'string'
-                                  || defaultType === 'number'
-                                  || defaultType === 'boolean';
-            }
-
-            if (isPrimitiveType) {
+            const isPrimitiveTypeInFastMode = checkIsPrimitiveTypeInFastMode(defaultValue, userType);
```
Comment (minggo) [inline cocos/serialization/deserialize-dynamic.ts]: Can i know why name it as fast mode?

Comment (shrinktofit) [reply cocos/serialization/deserialize-dynamic.ts]: You can ask @jareguo . I also documented the definition here: https://github.com/cocos/cocos-engine/pull/15630/commits/a5a2ccd3c2e2b378705478bba900cd3a6a542cc5 .

Comment (jareguo) [reply cocos/serialization/deserialize-dynamic.ts]: Yes, just as shrinktofit's remarks. FastMode does not verify every instance data that is loaded, but simply processes it uniformly. This is an optimization strategy that can only be used when the types of instance data are certain to match, and is currently only effective for types built into the engine. When the type of the engine really changes and needs to be compatible, the editor can complete the data migration, so all types of the engine are fast. However, the instance data in the user project may change continuously as the code is adjusted, so the loaded data needs to be checked twice.

File: cocos/serialization/deserialize-dynamic.ts [hunk 3]
```diff
@@ -281,6 +265,28 @@ function compileDeserializeNative (_self: _Deserializer, klass: CCClassConstruct
...
+ * In fast mode, deserialization of the class will go into an optimized way:
+ * each class property will be examined whether to be primitive according to their default value
+ * and type. Finally, all primitive properties would be together deserialized using simple assignment,
+ * without performing in-loop check.
+ */
+function canBeDeserializedInFastMode (klass: any): boolean {
+    return misc.BUILTIN_CLASSID_RE.test(js.getClassId(klass));
+}
+
+function checkIsPrimitiveTypeInFastMode (defaultValue: any, userType: any): boolean {
```
Comment (minggo) [inline cocos/serialization/deserialize-dynamic.ts]: As these two functions are check whether can treat them as fast mode, and TypeScript can not override function, so i suggest to rename them to
- canBeDeserializedInFastModeForClass
- canBeDeserializedInFastModeForType

File: cocos/serialization/deserialize-dynamic.ts (hunk 2 above)
Comment (minggo) [reply cocos/serialization/deserialize-dynamic.ts]: Good to know. Thanks.

File: cocos/serialization/deserialize-dynamic.ts (hunk 3 above)
Comment (shrinktofit) [reply cocos/serialization/deserialize-dynamic.ts]: > As these two functions are check whether can treat them as fast mode, and TypeScript can not override function, so i suggest to rename them to
> 
> * canBeDeserializedInFastModeForClass
> * canBeDeserializedInFastModeForType

No, the function you outlined is not what you expressed. It expresses "in fast mode, is a property with `defaultValue` and `userType`" is primitive?

Comment (minggo) [reply cocos/serialization/deserialize-dynamic.ts]: Then what does fast mode mean for the two functions?

Comment (minggo) [reply cocos/serialization/deserialize-dynamic.ts]: fast mode in two functions have different meaning?

**Review by minggo — APPROVED** [review]

**Pull request merged**